File: pythonPrototype/machinePlayer.py
import random

import numpy as np


class Player:
    def __init__(
        self,
        initial_coins: int,
        max_num_to_take: int,
        learning_rate: float = 0.5,
        discount_rate: float = 0.9,
    ) -> None:
        self.__initial_coins = initial_coins
        self.__max_num_to_take = max_num_to_take
        self.table = np.zeros((self.__initial_coins, self.__max_num_to_take))
        self.__path: list[tuple[int, int]] = []  # [(rest coin(s), take number), ...]
        self.__learning_rate = learning_rate
        self.__discount_rate = discount_rate

    def refresh_path(self) -> None:  # Refresh when game is over.
        self.__path.clear()

    # Receive feedback either when the game is over or when invalid move made.
    def receive_feedback(self, feedback: float) -> None:
        if feedback != float("-inf"):
            self.table[self.__path[-1][0] - 1, self.__path[-1][1] - 1] += feedback
        else:  # invalid move made
            prev_state, prev_num_took = self.__path.pop()

            # Revert the values that last invalid move added.
            self.__update_table(current_coin=prev_state, reverse=True)

            self.table[prev_state - 1, prev_num_took - 1] += feedback

    def __update_table(self, current_coin: int, reverse: bool = False) -> None:
        def get_score(take: int) -> float:
            coin_you_left = current_coin - take
            if coin_you_left in (0, 1):
                return 0.0
            elif coin_you_left > 1:
                max_coin_your_opponent_left = coin_you_left - 1
                next_possible_states_for_you = self.table[
                    (max_coin_your_opponent_left - 1) :: -1, ::
                ][: self.__max_num_to_take]

                # Your opponent is always smart enough to push you to the worst
                # or the most uncertain situation!
                max_score_of_each_row = np.max(next_possible_states_for_you, axis=1)
                return np.min(max_score_of_each_row) * self.__discount_rate
            else:
                raise ValueError

        for i in range(1, min(self.__max_num_to_take, current_coin) + 1):
            # Q <- Q + lr * score
            self.table[current_coin - 1, i - 1] += (
                self.__learning_rate * get_score(i) * (-1 if reverse else 1)
            )

    def make_move(self, current_coin: int) -> int:
        self.__update_table(current_coin)
        current_row: np.ndarray = self.table[current_coin - 1]
        if np.all(current_row == 0):
            take = random.randint(1, self.__max_num_to_take)
        else:
            take = int(np.argmax(current_row)) + 1
        self.__path.append((current_coin, take))
        return take


if __name__ == "__main__":
    ...

File: webpage/machinePlayer.ts
export class MachinePlayer {
    protected _maxPickable: number;
    private _table: number[][];
    // Below are how the table looks like...
    // -------------------------------------------------
    //              take1coin    take2coins  take3coins
    // 1coinRest
    // 2coinsRest
    // .
    // .
    // .
    // 10coinsRest
    // -------------------------------------------------
    private _path: number[][];
    private _lr: number;
    private _discountRate: number;
    public winTimes: number;

    constructor(initialNumOfCoin: number, maxPickable: number) {
        this._maxPickable = maxPickable;
        this._table = Array(initialNumOfCoin).fill(null).map(() => Array(maxPickable).fill(0));
        this._path = []; // [[rest coin(s), number took], ...]
        this._lr = 0.5;
        this._discountRate = 0.5;
        this.winTimes = 0;
    }
    public get table(): number[][] {
        return this._table;
    }
    private appendOnPath(coinRest: number, numTook: number): void {
        this._path.push([coinRest, numTook]);
    }
    public refreshPath(): void {
        this._path = [];
    }
    public receiveFeedback(feedBack: 1 | -1 | -2): void | number {
        const targetRowIdx: number = this._path[this._path.length - 1][0] - 1;
        const targetColumnIdx: number = this._path[this._path.length - 1][1] - 1;
        if (feedBack == 1) {    // win
            this._table[targetRowIdx][targetColumnIdx] += 10;
        } else if (feedBack == -1) {    // lose
            this._table[targetRowIdx][targetColumnIdx] -= 10;
        } else if (feedBack == -2) {    // if an invalid move has been made
            // make a deduction (reverse) of the values the latest invalid move added.
            // "targetRowIdx + 1" is the "latest number of coins rest"
            this.updateScoresOfARow(targetRowIdx + 1, true);
            this._table[targetRowIdx][targetColumnIdx] = -1 * Infinity;
            this._path.pop();
            return this.makeMove(targetRowIdx + 1);
        }
    }
    public makeMove(restCoin: number): number {
        // learn something before make choice
        this.updateScoresOfARow(restCoin, false);
        const targetRow: number[] = this._table[restCoin - 1];
        let numTake: number;
        let maxScore: number;
        if (targetRow.every((i) => i == 0)) {
            numTake = Math.floor(Math.random() * this._maxPickable) + 1;
            maxScore = targetRow[numTake - 1];
        } else {
            maxScore = targetRow[0];
            numTake = 1;
            for (let i = 0; i < targetRow.length; i++) {
                if (targetRow[i] > maxScore) {
                    maxScore = targetRow[i];
                    numTake = i + 1;
                }
            }
        }
        this.appendOnPath(restCoin, numTake)
        return numTake;
    }
    private updateScoresOfARow(restCoin: number, reverse: boolean): void {
        const sign = reverse ? -1 : 1;
        for (let i = 1; i <= this._maxPickable; i++) {
            if (i <= restCoin) {
                // Q <- Q + lr * point
                this._table[restCoin - 1][i - 1] += sign * this._lr * this.calcScore(restCoin, i);
            }
        }
    }
    private calcScore(restCoin: number, numTake: number): number {
        const restCoinIfYouTakeN = restCoin - numTake;
        // If, afterwards, your opponent take "m" coins,
        // you will heve "restCoinIfYouTakeN - m" coins rest the next time.
        if (restCoinIfYouTakeN > 1) {
            let concernPart: number[][] = [];
            for (let i = 0; i < Math.min(this._maxPickable, restCoinIfYouTakeN - 1); i++) {
                concernPart.push(this._table[(restCoinIfYouTakeN - 1) - 1 - i]);
            }

            // If there exists any row that all numbers are negative,
            // it means that "take i" now will lead you to failure
            // because your opponent is smart enough to push you to that situation.
            for (let eachRow of concernPart) {
                if (eachRow.every((i) => i < 0)) {
                    return Math.max(...eachRow) * this._discountRate;
                }
            }

            // If there exists any row that all numbers are zero
            // (and no row is all-negative) it means that this choice
            // will lead you to an unknown result.
            for (let eachRow of concernPart) {
                if (eachRow.every((i) => i == 0)) return 0;
            }

            // If non of the above two situations happened...
            let oneDArray: number[] = [];
            for (let eachRow of concernPart) {
                for (let eachScore of eachRow) oneDArray.push(eachScore);
            }
            return Math.max(...oneDArray) * this._discountRate;
        } else return 0;
    }
}

export class RandomPlayer extends MachinePlayer {
    public makeMove(restCoin: number): number {
        return Math.floor(Math.random() * Math.min(this._maxPickable, restCoin) + 1);
    }
    public receiveFeedback(feedBack: number): void | number { }
}